File: python/kth_largest_element.py
import heapq
import random
class Solution(object):
    def findKthLargest(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: int
        """
        # Set up a heap (python default min-heap),
        # so, negative to simulate a max-heap
        heap = [-x for x in nums]
        heapq.heapify(heap)
        # pop out the 1 to k-1 th element
        for i in xrange(1, k):
            heapq.heappop(heap)
        # Return the kth element
        return -heapq.heappop(heap)

class Solution2(object):
    """
    Method 2: Divide and Conquer
    """
    def findKthLargest(self, nums, k):
        return self.randomSelect(nums, k, 0, len(nums)-1)

    def partition(self, nums, p, r):
        i, j = p+1, r
        x = nums[p]  # pivot value
        while True:
            while nums[i] <= x and i < r:
                i += 1
            while nums[j] >= x and j > p:
                j -= 1
            if i >= j:
                break
            # swap, small to left
            nums[i], nums[j] = nums[j], nums[i]
        if i == r and j == p:
            return (p+r)/2
        # nums[j] <= x
        nums[p], nums[j] = nums[j], x
        return j

    def randomPartitoin(self, nums, p, r):
        # partition nums
        # s = random.randint(p, r)
        print "s=%d" %s
        nums[p], nums[s] = nums[s], nums[p]
        return self.partition(nums, p, r)

    def randomSelect(self, nums, k, p, r):
        # print "p=%d, r=%d" %(p, r)
        if p == r:
            return nums[p]
        i = self.randomPartitoin(nums, p, r)
        # print 'i=%d' %i
        n_small = i - p + 1  # number of left part
        if k > n_small:
            return self.randomSelect(nums, k-n_small, i+1, r)
        return self.randomSelect(nums, k, p, i)

if __name__ == '__main__':
    # s = Solution()
    # print s.findKthLargest([0,1,2,3,4,5,6], 2)
    s = Solution2()
    nums = [4,5,2,1,7,9]
    # s.partition(nums, 0, len(nums)-1)
    print s.findKthLargest(nums, 2)
    # print nums

File: cpp/kth_largest_element.cpp
#include <iostream>
#include <vector>
using namespace std;

class MaxHeap {
private:
    int heap_size;
    int parent(int i) {return (i-1)/2; }
    int left(int i)  {return 2*i +1; }
    int right(int i) {return 2*i +2; }
    void heapify(vector<int> &A, int i);

public:
    explicit MaxHeap(int size):
        heap_size(size)
        {}
    void build_heap(vector<int> &A);
    int heappop(vector<int> &A);
    int getHeapsize() {return heap_size; }
};

void MaxHeap::heapify(vector<int> & A, int i) {
    int l = left(i);
    int r = right(i);
    int largest = 0;
    if (l < heap_size && A[l] > A[i]){
        largest = l;
    }else {
        largest = i;
    }
    if (r < heap_size && A[r] > A[largest]) {
        largest = r;
    }
    if (largest != i) {
        std::swap(A[i], A[largest]);
        heapify(A, largest);
    }
}

void MaxHeap::build_heap(vector<int> &A) {
    for (int i=(A.size()-1)/2; i>=0; i--) {
        heapify(A, i);
    }
}

int MaxHeap::heappop(vector<int> &A) {
    cout<<"swap : "<<A[0]<<", "<<A[heap_size-1]<<endl;
    std::swap(A[0], A[(heap_size--)-1]);
    heapify(A, 0);
    return A[heap_size];
}

int main() {
    vector<int> A;
    for (int i=0; i<10; i++) {
        A.push_back(i);
    }
    MaxHeap heap(A.size());
    heap.build_heap(A);
    for (int i=0; i<A.size(); i++) {
//        cout<<A[i]<<"  ";
        cout<<"heap size : "<<heap.getHeapsize()<<endl;
        cout<<heap.heappop(A)<<endl;
        for (int j=0; j<A.size(); j++) {
            cout<<A[j]<<"  ";
        }
        cout<<endl<<endl;
    }
}
